tests/TestCase/ORM/QueryTest.php

```diff
@@ -3548,4 +3562,142 @@ public function testNotMatchingNested()
+     *
+     * @return void
+     */
+    public function testAllNoDuplicateTypeCasting()
+    {
+        $table = $this->getTableLocator()->get('Comments');
+        $query = $table->find()
+            ->select(['id', 'comment', 'created']);
+
+        // Convert to an array and make the query dirty again.
+        $result = $query->all()->toArray();
+        $query->limit(99);
+
+        // Get results a second time.
+        $result2 = $query->all()->toArray();
+
+        $this->assertEquals(1, $query->__debugInfo()['decorators'], 'Only one typecaster should exist');
+    }
+
+    /**
+     * Test to see that the excluded fields are not in the select clause
+     *
+     * @return void
+     */
+    public function testSelectAllExcept()
+    {
+        $table = $this->getTableLocator()->get('Articles');
+        $result = $table
+            ->find()
+            ->selectAllExcept($table, ['body']);
+        $selectedFields = $result->clause('select');
+        $expected = [
+            'Articles__id' => 'Articles.id',
+            'Articles__author_id' => 'Articles.author_id',
+            'Articles__title' => 'Articles.title',
+            'Articles__published' => 'Articles.published'
+        ];
+        $this->assertEquals($expected, $selectedFields);
+    }
+
+    /**
+     * Test that the excluded fields are not included
+     * in the final query result.
+     *
+     * @return void
+     */
+    public function testSelectAllExceptWithContains()
+    {
+        $table = $this->getTableLocator()->get('Articles');
+        $this->getTableLocator()->get('Articles');
+        $table->hasMany('Comments');
+        $table->belongsTo('Authors');
+
+        $result = $table
+            ->find()
+            ->contain([
+                'Comments' => function ($query) use ($table) {
+                    return $query->selectAllExcept($table->Comments, ['published']);
+                }
+            ])
+            ->selectAllExcept($table, ['body'])
+            ->first();
+        $this->assertNull($result->comments[0]->published);
+        $this->assertNull($result->body);
+        $this->assertNotEmpty($result->id);
+        $this->assertNotEmpty($result->comments[0]->id);
+    }
+
+    /**
+     * Test what happens if you call selectAllExcept() more
+     * than once.
+     *
+     * @return void
+     */
+    public function testSelectAllExceptWithMulitpleCalls()
+    {
+        $table = $this->getTableLocator()->get('Articles');
+        $this->getTableLocator()->get('Articles');
+
+        $result = $table
+            ->find()
+            ->selectAllExcept($table, ['body'])
+            ->selectAllExcept($table, ['published']);
+        $selectedFields = $result->clause('select');
+        $expected = [
+            'Articles__id' => 'Articles.id',
+            'Articles__author_id' => 'Articles.author_id',
+            'Articles__title' => 'Articles.title',
+            'Articles__published' => 'Articles.published',
```

Comment:
Shouldn't published be excluded as it was in the `selectAllExcept` call on 3650?

Reply:
These tests show what would happen on multiple calls. As far as I know things go into $this->_parts but don't come out unless you reset/clear it. So on the first call all the columns but the excluded list go into the query. On a second call you can't then remove something from the query parts without resetting. So these tests show that behavior. This test also shows the "reset" behavior. Does that answer your question?

Reply:
Ok. Could we document how multiple calls interact in the API doc block? Having that should help people understand the method better.

Reply:
Will do. Thanks for the feedback.